DOC/mkproblem.1
.\" Automatically generated by Pod::Man v1.34, Pod::Parser v1.13
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MKPROBLEM 1"
.TH MKPROBLEM 1 "2003-04-29" "perl v5.8.0" "User Contributed Perl Documentation"
.SH "NAME"
mkproblem.pl ( Problem generating script for MPTP)
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
mkproblem.pl [options] problemnames
.PP
mkproblem.pl \-tcard_1 \-trolle \-ccard_2 t39_absvalue by_25_16_2_absvalue
.PP
.Vb 10
\& Options:
\&   --skipbadrefs[=<arg>],   -s[<arg>]
\&   --allownonex[=<arg>],    -a[<arg>]
\&   --basedir=<arg>,         -b<arg>
\&   --memlimit=<arg>,        -M<arg>
\&   --tharticles=<arg>,      -t<arg>
\&   --chkarticles=<arg>,     -c<arg>
\&   --filter=<arg>,          -f<arg>
\&   --help,                  -h
\&   --man
.Ve
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-\-skipbadrefs[=<arg>], \-s[<arg>]\fR" 8
.IX Item "--skipbadrefs[=<arg>], -s[<arg>]"
Skip problems with bad references.
.IP "\fB\-\-allownonex[=<arg>], \-a[<arg>]\fR" 8
.IX Item "--allownonex[=<arg>], -a[<arg>]"
Setting to 0 causes error exit on nonexistant articles.
.IP "\fB\-\-basedir=<arg>, \-b<arg>\fR" 8
.IX Item "--basedir=<arg>, -b<arg>"
Sets the \s-1MPTPDIR\s0 to <arg>.
.IP "\fB\-\-memlimit=<arg>, \-M<arg>\fR" 8
.IX Item "--memlimit=<arg>, -M<arg>"
Sets memory limit for database caches, not implemented yet.
.IP "\fB\-\-tharticles=<arg>, \-t<arg>\fR" 8
.IX Item "--tharticles=<arg>, -t<arg>"
Do all theorem problems from the article <arg>, this option can be 
repeated to specify multiple articles.
.IP "\fB\-\-chkarticles=<arg>, \-c<arg>\fR" 8
.IX Item "--chkarticles=<arg>, -c<arg>"
Do all checker problems from the article <arg>, this option can be 
repeated to specify multiple articles.
.IP "\fB\-\-filter=<arg>, \-f<arg>\fR" 8
.IX Item "--filter=<arg>, -f<arg>"
Specify the filtering of the background formulas. 
Default is now 1 \- the checker-based signature filtering. 
Setting to 0 does no filtering at all.
.IP "\fB\-\-help, \-h\fR" 8
.IX Item "--help, -h"
Print a brief help message and exit.
.IP "\fB\-\-man\fR" 8
.IX Item "--man"
Print the manual page and exit.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBmkproblem.pl\fR gets a list of options and a list of theorem names or
checker problem names (in our format, e.g. t39_absvalue or
by_25_16_2_absvalue), and produces a complete dfg file 
for each of them. This file contains full informations 
necessary for reproval, i.e. the references and all of the background
info (constructor types, requirements, etc.).
Problem files are stored under their articles directories,
in the \s-1PROBLEMS\s0 directory.
.SH "CONTACT"
.IX Header "CONTACT"
Josef Urban urban@kti.ms.mff.cuni.cz


SCRIPT/mkproblem.pl
#!/usr/bin/perl -w

=head1 NAME

mkproblem.pl ( Problem generating script for MPTP)

=head1 SYNOPSIS

mkproblem.pl [options] problemnames

mkproblem.pl -tcard_1 -trolle -ccard_2 t39_absvalue by_25_16_2_absvalue

 Options:
   --skipbadrefs[=<arg>],   -s[<arg>]
   --allownonex[=<arg>],    -a[<arg>]
   --basedir=<arg>,         -b<arg>
   --memlimit=<arg>,        -M<arg>
   --tharticles=<arg>,      -t<arg>
   --chkarticles=<arg>,     -c<arg>
   --filter=<arg>,          -f<arg>
   --definitions=<arg>,     -D<arg>
   --specfile=<arg>,        -F<arg>
   --help,                  -h
   --man

=head1 OPTIONS

=over 8

=item B<<< --skipbadrefs[=<arg>], -s[<arg>] >>>

Skip problems with bad references.

=item B<<< --allownonex[=<arg>], -a[<arg>] >>>

Setting to 0 causes error exit on nonexistant articles.

=item B<<< --basedir=<arg>, -b<arg> >>>

Sets the MPTPDIR to <arg>.

=item B<<< --memlimit=<arg>, -ME<lt>arg> >>>

Sets memory limit for database caches, not implemented yet.

=item B<<< --tharticles=<arg>, -t<arg> >>>

Do all theorem problems from the article <arg>, this option can be 
repeated to specify multiple articles.

=item B<<< --chkarticles=<arg>, -c<arg> >>>

Do all checker problems from the article <arg>, this option can be 
repeated to specify multiple articles.

=item B<<< --filter=<arg>, -f<arg> >>>

Specify the filtering of the background formulas. 
Default is now 1 - the checker-based signature filtering. 
Setting to 0 does no filtering at all.

=item B<<< --definitions=<arg>, -B<D><arg> >>>

Specify treatment of definitions.
Default is now 0 - the 'definitions' directive is neglected
and definitions must be explicitely specified as references
to be included as axioms for the problems.
1 means that all definitions from the 'definitions' directive
are included as direct references.
2 is as 1, but the definitions are treated as background formulas
instead, i.e. if filtering is activated, they are included only
if the defined symbols appear during the fixpoint computation.
3 is like 1, but only definitions from the current article
are used.
4 is like 2, again with current article definitions only.
Options 3 and 4 are useful when we know that the definitions from
other articles have been taken care of explicitly (e.g. by some
previous experience like Mizar Proof Advisor), but this does not
extend to the current article, which may be new in some sense.

=item B<<< --specfile=<arg>, -B<F><arg> >>>

Read problem specifications from the file <arg>.

=item B<<< --help, -h >>>

Print a brief help message and exit.

=item B<<< --man >>>

Print the manual page and exit.

=back

=head1 DESCRIPTION

B<mkproblem.pl> gets a list of options and a list of theorem names or
checker problem names (in our format, e.g. t39_absvalue or
by_25_16_2_absvalue), and produces a complete dfg file 
for each of them. This file contains full informations 
necessary for reproval, i.e. the references and all of the background
info (constructor types, requirements, etc.).
Problem files are stored under their articles directories,
in the PROBLEMS directory.

=head1 CONTACT

Josef Urban urban@kti.ms.mff.cuni.cz

=cut

#------------------------------------------------------------------------
#
# File  : mkproblem.pl 
#
# Author: Josef Urban
#
# Gets a list of options and a list of theorem names or
# checker problem names (in our format, e.g. "t39_absvalue" or
# "by_25_16_absvalue"), and produces
# a complete dfg file for each of them. This file
# contains full informations necessary for reproval,
# i.e. the references and all of the background
# info (constructor types, requirements, etc.).
#
# Problem files are stored under their articles' directories.
#
# Changes
#
# <1> Tue Feb 11 21:26:44 2003
#     New
#------------------------------------------------------------------------
use strict;
use Getopt::Long;
use Pod::Usage;
use MPTPDebug;
use MPTPUtils;
use MPTPSgnFilter;


our (
     %gproblems
     #     %th_problems, 
     #     %chk_problems
    );


# %gproblems has for each article following slots:
# 'BG'  - background theory
# 'THE' - theoem problems
# 'CHK' - checker problems
# 'DEF' - definitions moved here from 'BG', if $DEFINITIONS is of some REFS kind
undef %gproblems;
my ($help, $man, @tharticles, @chkarticles); # do all problems for these


my $gaddme = 1;			# Tells to add article to its env. directives
my $FILTER = 1;
my $DEFINITIONS = 0;
my $ProblemSpecFile;            # Input file containing problem specifications

sub Usage 
{
    print "Need a list of problem names like 't39_absvalue' for 
theorem problems and 'by_25_16_2_absvalue' for checker problems.\nUse the -t option (e.g. -tcard_1 -treal_1) to do all theorems for an article, and the -c option to do all checker problems for an article.\nUse the -b option to override the MPTPDIR variable.";
    exit;
}


#------------------------------------------------------------------------
#  Function    : ProcessArgs()
#
#  Check the arguments - article names and theorem numbers.
#
#  Input       : -
#  Global Vars : %gcnt, @ARGV, %gproblems, $ProblemSpecFile
#  Side Effects: Can die, loads %gproblems
#------------------------------------------------------------------------

sub  ProcessArgs
{
    my ($arg, $thnr, $an);

    if(defined $ProblemSpecFile)
    {
	open(SPECFILE, $ProblemSpecFile) or
	    die "$ProblemSpecFile not readable!";

	while (<SPECFILE>) { ProcessArg($_); }
	close(SPECFILE);
    }

    foreach $arg (@ARGV) { ProcessArg($arg); }
}

#------------------------------------------------------------------------
#  Function    : ProcessArg()
#
#  Check the arguments - article names and theorem numbers.
#
#  Input       : -
#  Global Vars : %gcnt, @ARGV, %gproblems, 
#  Side Effects: Can die, loads %gproblems
#------------------------------------------------------------------------

sub  ProcessArg
{
    my ($arg) = @_;
    my ($thnr, $an);

    if ($arg =~ /t(\d+)_(\w+)(\[.*\])?$/)
    {
	$thnr = $1;
	$an   = $2;

	exists $gcnt{$an} or 
	    die "Article $an unknown, add it into db!";
	$thnr <= $gcnt{$an}->{'THE'} or 
	    die "$thnr is greater than the max. theorem number for $an";

	$gproblems{$an}{'THE'}{$thnr} = (defined $3) ? $3 : ();
    }

    elsif ($arg =~ /by_(\d+)_(\d+)_(\d+)_(\w+)/)
    {
	exists $gcnt{$4} or 
	    die "Article $3 unknown, add it into db!";
	$3 <= $gcnt{$4}->{'CHK'} or 
	    die "$3 is greater than the max. checker problem number for $an";
	$gproblems{$4}{'CHK'}{$3} = $arg;
    }

    else
    {
	Usage();
    }
}



sub AddWholeArticles
{
    my ($an, $i);

    foreach $an (@chkarticles)
    {
	exists $gcnt{$an} or 
	    die "Article $an unknown, add it into db!";

	for ($i = 1; $i <= $gcnt{$an}->{'CHK'}; $i++)
	{
	    $gproblems{$an}{'CHK'}{$i} = ()
		unless (exists $gproblems{$an}{'CHK'}{$i});
	}
    }


    foreach $an (@tharticles)
    {
	exists $gcnt{$an} or 
	    die "Article $an unknown, add it into db!";

	for ($i = 1; $i <= $gcnt{$an}->{'THE'}; $i++)
	{
	    $gproblems{$an}{'THE'}{$i} = ()
		unless (ThCanceled($i + $grcn{$an}->{'THE'} - 1));
	}
    }

}



#------------------------------------------------------------------------
#  Function    : CreateDirs()
#
#  Create the directories in %gproblems, probably can set %errno,
#  no checking now.
#
#  Input       : -
#  Global Vars : %gproblems
#  Side Effects: I/O
#------------------------------------------------------------------------

sub CreateDirs
{
    my $dir;

    foreach $dir (keys %gproblems)
    {
	mkdir $MPTPPROBLEMS.$dir;
    }
}


sub PrepareBgs
{
    my $key;

    foreach $key (keys %gproblems)
    {
	$gproblems{$key}{'BG'} = CreateBg($key, $gaddme, $DEFINITIONS);


	if(($DEFINITIONS == DEFS_ALL_AS_REFS) 
	   || ($DEFINITIONS == DEFS_SELF_AS_REFS))
	{
	    $gproblems{$key}{'DEF'} = $gproblems{$key}{'BG'}->{'DEF'};
	    delete $gproblems{$key}{'BG'}->{'DEF'};
	}
    }
}

#------------------------------------------------------------------------
#  Function    : DoProblems()
#
#  Create the problems. The existence check is supposed
#  to be done before calling this. 
#
#  Input       : problem references, article background theory
#  Global Vars : %gcnt, %GNTOKENS
#  Output      : filtered background theory
#  Side Effects: I/O
#------------------------------------------------------------------------

sub DoProblems
{
    my ($an,$nr,$prb,%refnbrs,$ref,$myfname,$chkrefs,
	$mybg,$mysymbols,$dirkind,$j,%bgcache,%bgsyms,$pname);

    foreach $an (sort {$gcnt{$a}->{'ORD'} <=> $gcnt{$b}->{'ORD'}} 
		 keys %gproblems)
    {
	undef %bgcache;
	undef %bgsyms;

	if ($FILTER == 0)
	{
	    $mybg      = $gproblems{$an}{'BG'};
	    GetAllBgSyms($mybg, \%bgsyms);     # OK for any kind of $DEFINITIONS
	}

    TH: foreach $nr ( sort {$a <=> $b} (keys %{$gproblems{$an}{'THE'}}) )
	{
	    %refnbrs = GetThRefs($an, $nr, $gproblems{$an}{'THE'}{$nr});

	    if(exists $gproblems{$an}{'DEF'})  # Add the DEFS if as REFS
	    {
		my $value;
		my %tmp_hash = ();

		@tmp_hash{ @{$refnbrs{'DEF'}} } = ()
		    if(exists $refnbrs{'DEF'});

		foreach $value (@{ $gproblems{$an}{'DEF'}})
		{
		    (exists $tmp_hash{$value}) or push( @{$refnbrs{'DEF'}}, $value);
                }
	    }

	    if (! (exists $refnbrs{'THE'}))
	    {
		print "Problem t$nr\_$an ignored, unexported references\n"
		    if (GWATCHED & WATCH_BADREFS);

		next TH;
	    }

	    if ($FILTER == 0)  # OK for any kind of $DEFINITIONS
	    {
		$mysymbols = AddSymsAndSpecials('THE', \%refnbrs,
						$mybg, \%bgsyms);
	    }
	    else
	    {
		($mybg, $mysymbols) = FilterBg('THE', \%refnbrs,
					       $gproblems{$an}{'BG'}, \%bgcache);
		$mysymbols = GroupDfgSymbols($mysymbols);
	    }

	    $myfname = $MPTPPROBLEMS.$an."/t".$nr."_".$an.".dfg";

	    PrintDfgProblem('THE', $myfname, "t".$nr."_".$an, 
			    $mysymbols, $mybg, \%refnbrs);

	    undef $mybg->{'NDB'} if ($FILTER == 0);
	}

	if ( exists $gproblems{$an}{'CHK'} )
	{
	    OpenChkDb($an);

	CHK: foreach $prb (sort {$a<=>$b} keys %{$gproblems{$an}{'CHK'}})
	    {
		#	      $prb =~ /by_(\d+)_(\d+)_(\d+)_(\w+)/;
		($pname, $chkrefs)  = GetChkRefs($an, $prb);

		# assert($prb eq $pname);

		if ($FILTER == 0)
		{
		    $mysymbols = AddSymsAndSpecials('CHK', $chkrefs,
						    $mybg, \%bgsyms);
		}
		else
		{

		    ($mybg, $mysymbols)	 =
			FilterBg('CHK', $chkrefs,
				 $gproblems{$an}{'BG'}, \%bgcache);
		    $mysymbols		 = GroupDfgSymbols($mysymbols);
		}

		$myfname = $MPTPPROBLEMS.$an."/".$pname.".dfg";

		PrintDfgProblem('CHK', $myfname, $pname, 
				$mysymbols, $mybg, $chkrefs);

		undef $mybg->{'NDB'} if ($FILTER == 0);
	    }

	    CloseChkDb($an);
	}



    }
}



Getopt::Long::Configure ("bundling","no_ignore_case");

GetOptions('skipbadrefs|s:i'   => \$SkipBadThRefsProblems,
	   'allownonex|a:i'    => \$AllowNonExistant,
	   'basedir|b=s'     => \$MPTPDIR,
	   'memlimit|M=i'    => \$DBMEMLIMIT,
	   'tharticles|t=s'  => \@tharticles,
	   'chkarticles|c=s' => \@chkarticles,
	   'filter|f=i'      => \$FILTER,
	   'definitions|D=i' => \$DEFINITIONS,
	   'specfile|F=s'    => \$ProblemSpecFile,
	   'help|h'          => \$help,
	   'man'             => \$man)
    or pod2usage(2);

pod2usage(1) if($help);
pod2usage(-exitstatus => 0, -verbose => 2) if($man);
pod2usage(2) 
    if (($#ARGV < 0) && !(defined $ProblemSpecFile) 
	&& ($#tharticles < 0) && ($#chkarticles < 0)); 

pod2usage(2) if($DEFINITIONS > DEFS_SELF_AS_BG);

die "Set the MPTPDIR shell variable, or run with the -b option"
    if ("/" eq $MPTPDIR);


LoadCounts();

PrintCnts() if(GWATCHED & WATCH_COUNTS);
PrintRcns() if(GWATCHED & WATCH_COUNTS);

ProcessArgs();
OpenDbs();
AddWholeArticles();

# print (keys %gproblems);
CreateDirs();

PrepareBgs();

DoProblems();

__END__
